### ISSUE TITLE
Secrets mounting limited by kubernetes client throttling

### ISSUE THREAD

**Issue by dindurthy** [aws/secrets-store-csi-driver-provider-aws#136]
**Describe the bug**
We routinely see errors like in the secrets store driver
```
"failed to mount secrets store object content\" err=\"rpc error: code = Unknown desc = rate: Wait(n=1) would exceed context deadline\" pod=\"<some-pod-name>\""
```
And in many cases we see corresponding errors in csi aws provider:
```
Failed to initialize AWS session\" err=\"rate: Wait(n=1) would exceed context deadline\"
```

The latter is coming from https://github.com/aws/secrets-store-csi-driver-provider-aws/blob/c02c72deeebe81ae44cc00d9b931043f30455b82/server/server.go#L130.

The issue is I don't know what that error means. It looks to me like client-side throttling, but I'm unsure what the fundamental error is. I'm guessing that there is actual AWS server-side rate-limiting going on, but I'm not finding `ThrottlingException` or `Rate exceeded` or something definitive, and [it is notoriously a pain to work backwards from CloudTrail](https://aws.amazon.com/premiumsupport/knowledge-center/cloudtrail-rate-exceeded/).

This typically happens in the vicinity of upgrading services or draining nodes/scheduling onto new nodes, and it is node-specific. The node itself has free CPU and memory while these errors occur. 

**To Reproduce**

We tend to find this when draining a node of ~50 pods, each with 3-10 secret mounts.

Do you also notice this bug when using a different secrets store provider (Vault/Azure/GCP...)? **No**

If yes, the issue is likely with the k8s Secrets Store CSI driver, not the AWS provider. Open an issue in [that repo](https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues).

**Expected behavior**

`Failed to initialize AWS session` reports a meaningful error or some contingent error is separately logged.

**Environment:**
OS, Go version, etc.

**Additional context**

**Comment by dindurthy** [comment]
Also when I look at the csi-driver-provider logs for `Servicing mount request for pod <some-pod-name>`, I don't see evidence of an exponential backoff
![Screen Shot 2022-10-05 at 4 46 22 PM](https://user-images.githubusercontent.com/14302694/194183608-6cc4d2bb-3cfa-43d7-957b-4b9d192dd691.png)
 Maybe that would be the secrets-store-driver's responsibility.

**Comment by goyalya** [comment]
Thank you for reporting the issue. We are looking into the request.

**Comment by dindurthy** [comment]
One more observation in my testing -- it appears something seems to go wrong when the node is under high load, which happens when many pods are scheduled at the same time.
![Screen Shot 2022-10-07 at 2 06 21 PM](https://user-images.githubusercontent.com/14302694/194652465-150314d9-5c4f-44ff-977c-72065823472f.png)
The gap corresponds to the node being pegged. As the resources recover, pods stuck pending on the node continuously fail due to the aforemented errors indefinitely. Deleting those pods or restarting the csi-driver-provider-aws resolves the scheduling issue. The latter peaks in the graph correspond to my restarting the driver and many pending pods successfully lifting all at once.

So, it appears like the AWS session-per-pod in the csi-driver-provider-aws is broken indefinitely when pods are scheduled on a node with high load.

**Comment by goyalya** [comment]
@dindurthy : The recent example you shared seems to be separate issue from the original exception handling issue.  If you are still experiencing this issue please open a support case with AWS.

**Comment by dindurthy** [comment]
Ok, I was able to make some progress and I believe the issue is kubernetes go client rate limiting.

I upgraded to [latest](https://github.com/aws/secrets-store-csi-driver-provider-aws) which didn't solve the problem but improved the logging.

During an application upgrade or draining other nodes, we may schedule ~20-40 pods on a node and many of those pods have up to 30 secrets mounts/pod. The `csi-secrets-store-provider-aws` pod hits client rate limiting and pods become "stuck". 

For a while, initially, we see these occasional errors:

```
Waited for 16.996690473s due to client-side throttling, not priority and fairness, request: POST:https://10.100.0.1:443/api/v1/namespaces/services/serviceaccounts/redoxengine-default/token",
```

at a rate of 0-10/minute. This looks normal. Some pods successfully lift, but at some point, pending pods get stuck. All pods emit these errors:

```
\"failed to mount secrets store object content\" err=\"rpc error: code = Unknown desc = us-east-1: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline\" pod=\"services/redoxengine-cardinal-api-priority-674d4b5d5-thskl\"",
```
These are all definitely kubernetes go cli errors, e.g. [this](https://github.com/kubernetes/client-go/blob/228b004f4a926d43f53c0f8b6de3e1e2a3e50fcb/rest/request.go#L618) and [this](https://github.com/kubernetes/client-go/blob/228b004f4a926d43f53c0f8b6de3e1e2a3e50fcb/rest/request.go#L609).

At this point, pods are stuck indefinitely. Sometimes they lift after 20-30 minutes -- not always -- but typically we intervene or rollback. I examined the API actions that are happening by the csi pod. It looks like we are doing 3 kinds of lookups:

* [1 node lookup/pod/mount attempt](https://github.com/aws/secrets-store-csi-driver-provider-aws/blob/f67d9277c3b114a6e0d2cbc5041d7ea02144b245/server/server.go#L262)
* [1 pod lookup/pod/mount attempt](https://github.com/aws/secrets-store-csi-driver-provider-aws/blob/f67d9277c3b114a6e0d2cbc5041d7ea02144b245/server/server.go#L255)
* [1 service account token create/pod/mount attempt](https://github.com/aws/secrets-store-csi-driver-provider-aws/blob/f67d9277c3b114a6e0d2cbc5041d7ea02144b245/auth/auth.go#L52)

There may be more that I am missing. Unsurprisingly, each of these calls occur per Mount execution. It is not very efficient and there is no caching, and all of these API calls are necessary for the pods to schedule. So I believe we are in a storm of pods that attempt to node-lookup, pod lookup, token create, error, get throttled, and eventually the wait on the throttle would exceed the context deadline for these calls so they don't actually wait -- they just cancel -- and no mount attempt can actually succeed in order to relieve the pressure.

Mounting many classes/pod is not scalable. Seems like there could be a lot done to make this more efficient. I would be good to document some of these considerations somewhere -- maybe they are documented and I missed it. Minimizing the secrets/pod I think is our only option without significant enhancements. Possibly exposing kubernetes cli options could give us something to work with. But some kind of caching seems like the only option long term.

**Comment by jbct** [comment]
Hi @dindurthy. I know it's been a while since this was reported, but we recently discovered another customer case where client-side rate limiting was being hit. There is a default QPS limit of 5.0 and Burst of 10 in the Kubernetes Go client which is used. This can be changed by adding `cfg.QPS = 20.0` and `cfg.Burst = 40` (adjusting for whatever numbers you feel work for your environment) above the line in [this file](https://github.com/aws/secrets-store-csi-driver-provider-aws/blob/1aace58f2a6e8a003e99d340cad36dd3b4e24a07/main.go#L61). We have added a backlog item to make this configurable for customers in the future.
